File: Experiment Processing/recommender/distance_metrics/euclidean_distance.py
import math
import numpy as np

from recommender.distance_metrics.abstract_distance_metric import AbstractDistanceMetric


class EuclideanDistance(AbstractDistanceMetric):
    """
    Euclidean distance metric, see https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.euclidean.html
    for more information.
    """

    def get_name(self):
        return "Euclidean Distance"

    def calculate_distance(self, track_a, track_b):

        temp_distance = 0

        track_a_features = track_a.get_metadata()
        track_b_features = track_b.get_metadata()

        for attribute in self.attribute_keys:
            attribute_a = track_a_features[attribute]
            attribute_b = track_b_features[attribute]

            temp_distance += (attribute_a - attribute_b) ** 2

        try:
            return 1 - math.sqrt(temp_distance)
        except ZeroDivisionError:
            return np.NaN

File: Web Survey/resources/js/recommender/distance_metrics/EuclideanDistance.js
import AbstractDistanceMetric from "./AbstractDistanceMetric";

export default class EuclideanDistance extends AbstractDistanceMetric {

    constructor(props) {
        super(props);
        this.ownTrackRating = 1.0;
    }


    static getMetricName() {
        return "Euclidean Distance";
    }

    calculateDistance(track, otherTrack, attributes) {
        const self = this;

        let tempDistance = 0;

        this.attributeKeys.forEach(attribute => {

            let attributeOwn;
            let attributeOther;

            if(attribute === "loudness") {
                attributeOwn = self.fixLoudness(attributes[track][attribute]);
                attributeOther = self.fixLoudness(attributes[otherTrack][attribute])
            } else {
                attributeOwn = attributes[track][attribute];
                attributeOther = attributes[otherTrack][attribute];
            }

            tempDistance += (attributeOwn - attributeOther) ** 2;
        });
        let distance;

        try {
            distance = Math.sqrt(tempDistance);
        } catch(err){
            console.log("Zero division");
            return NaN;
        }
        return (1 - distance);
    }

}